### PR TITLE
feat: Sveltekit templates

### PR DESCRIPTION
**Pull request by melMass**
### What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Docs
- [ ] New Binding issue #___
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [x] Other, please describe:
	- [x] Template

### Does this PR introduce a breaking change?

- [ ] Yes, and the changes were approved in issue #___
- [x] No

### Checklist
- [x] When resolving issues, they are referenced in the PR's title (e.g `fix: remove a typo, closes #___, #___`)
- [ ] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/create-tauri-app/blob/dev/.changes/readme.md).
- [x] I have added a convincing reason for adding this feature, if necessary

### Other information

This is a WIP that I will finish either today or tomorrow:

- [x] Add the missing rust side (enums etc..) as per [this](https://github.com/tauri-apps/create-tauri-app/blob/dev/.github/CONTRIBUTING.md#templates)
- [x] Add the Svelte-Kit TS template

## Question for Maintainers (cross asked on Discord)

- Can the templates have a splash screen? In the case of Svelte-Kit the approach is a bit different so it might be interesting.. Or it's maybe more suited for an example than a template maybe.

**Answered**: 
> No, templates should be kept as simple as possible and all match oob

### CODE REVIEW

File: .changes/svelte-kit-templates.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+"create-tauri-app": minor
+"create-tauri-app-js": minor
+---
+
+Add `svelte-kit` template.
```
Comment (amrbashir) [inline .changes/svelte-kit-templates.md]:
```suggestion
Add `svelte-kit` and `svelte-kit-ts` template.
```

File: packages/cli/fragments/fragment-svelte-kit-ts/src/app.d.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+// See https://kit.svelte.dev/docs/types#app
+// for information about these interfaces
+// and what to do when importing types
+declare namespace App {
+  // interface Locals {}
+  // interface PageData {}
+  // interface Error {}
+  // interface Platform {}
```
Comment (amrbashir) [inline packages/cli/fragments/fragment-svelte-kit-ts/src/app.d.ts]: delete these comments

File: packages/cli/fragments/fragment-svelte-kit-ts/svelte.config.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import adapter from "@sveltejs/adapter-static";
```
Comment (amrbashir) [inline packages/cli/fragments/fragment-svelte-kit-ts/svelte.config.js]:
```suggestion
import staticAdapter from "@sveltejs/adapter-static";
```

File: packages/cli/fragments/fragment-svelte-kit/src/app.d.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+// See https://kit.svelte.dev/docs/types#app
+// for information about these interfaces
+// and what to do when importing types
+declare namespace App {
+  // interface Locals {}
+  // interface PageData {}
+  // interface Error {}
+  // interface Platform {}
```
Comment (amrbashir) [inline packages/cli/fragments/fragment-svelte-kit/src/app.d.ts]: delete

File: packages/cli/fragments/fragment-svelte-kit/svelte.config.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import adapter from "@sveltejs/adapter-static";
```
Comment (amrbashir) [inline packages/cli/fragments/fragment-svelte-kit/svelte.config.js]:
```suggestion
import staticAdapter from "@sveltejs/adapter-static";
```

File: packages/cli/src/template.rs [hunk 6]
```diff
@@ -23,6 +23,8 @@ pub enum Template {
     VueTs,
     Svelte,
     SvelteTs,
+    SvelteKit,
+    SvelteKitTs,
```
Comment (amrbashir) [inline packages/cli/src/template.rs]: Move to the end of the list here and in every list of templates in the repo.

**Review by amrbashir — CHANGES_REQUESTED** [review]
Thank you for your contribution. Please move `favicon.png` file to the `_assets_` folder and include it in the `_cta_manifest_` file

File: packages/cli/fragments/fragment-svelte-kit-ts/src/app.d.ts [hunk 7]
```diff
@@ -0,0 +1,6 @@
+// See https://kit.svelte.dev/docs/types#app
+// for information about these interfaces
+// and what to do when importing types
+declare namespace App {
+
+}
```
Comment (amrbashir) [inline packages/cli/fragments/fragment-svelte-kit-ts/src/app.d.ts]: nit
```suggestion
declare namespace App {}
```

File: packages/cli/src/package_manager.rs [hunk 8]
```diff
@@ -41,6 +41,8 @@ impl PackageManager {
                 Template::VueTs,
                 Template::Svelte,
                 Template::SvelteTs,
+                Template::SvelteKit,
+                Template::SvelteKitTs,
```
Comment (amrbashir) [inline packages/cli/src/package_manager.rs]: nit: move to the end of the list

File: packages/cli/fragments/fragment-svelte-kit/src/app.d.ts [hunk 9]
```diff
@@ -0,0 +1,5 @@
+// See https://kit.svelte.dev/docs/types#app
+// for information about these interfaces
+// and what to do when importing types
+declare namespace App {
+}
```
Comment (amrbashir) [inline packages/cli/fragments/fragment-svelte-kit/src/app.d.ts]: nit
```suggestion
declare namespace App {}
```

**Review by amrbashir — CHANGES_REQUESTED** [review]

**Comment by amrbashir** [comment]
run `pnpm format` and `cargo fmt --all` before commiting

File: packages/cli/fragments/fragment-svelte-kit-ts/src/routes/+page.svelte [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+<h1>Welcome to Tauri!</h1>
+
+<div class="row">
+  <a href="https://vitejs.dev" target="_blank">
+    <img src="/vite.svg" class="logo vite" alt="Vite Logo" />
+  </a>
+  <a href="https://tauri.app" target="_blank">
+    <img src="/tauri.svg" class="logo tauri" alt="Tauri Logo" />
+  </a>
+  <a href="https://svelte.dev" target="_blank">
```
Comment (amrbashir) [inline packages/cli/fragments/fragment-svelte-kit-ts/src/routes/+page.svelte]:
```suggestion
  <a href="https://kit.svelte.dev" target="_blank">
```

File: packages/cli/fragments/fragment-svelte-kit/src/routes/+page.svelte [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+<h1>Welcome to Tauri!</h1>
+
+<div class="row">
+  <a href="https://vitejs.dev" target="_blank">
+    <img src="/vite.svg" class="logo vite" alt="Vite Logo" />
+  </a>
+  <a href="https://tauri.app" target="_blank">
+    <img src="/tauri.svg" class="logo tauri" alt="Tauri Logo" />
+  </a>
+  <a href="https://svelte.dev" target="_blank">
```
Comment (amrbashir) [inline packages/cli/fragments/fragment-svelte-kit/src/routes/+page.svelte]:
```suggestion
  <a href="https://kit.svelte.dev" target="_blank">
```

**Review by amrbashir — CHANGES_REQUESTED** [review]

**Comment by melMass** [comment]
@amrbashir Done the edits, the only thing I find strange is splitting them in the CLI options no? (sveltekit at the end of the list)
image

**Comment by amrbashir** [comment]
Templates are always added to the bottom of the list.

**Review by amrbashir — APPROVED** [review]
Thank you

**Comment by JDR42** [comment]
@melMass This is great to get SvelteKit in the mix, thank you!  However, the template might need a fix.  Using `appWindow` throws `ReferenceError: window is not defined` (when running dev server, i.e., `tauri dev`).

For example:
```
//  src/routes/+page.svelte

import { appWindow } from '@tauri-apps/api/window';
...
await appWindow.show();
...
```

Error:
```
window is not defined
ReferenceError: window is not defined
    at Object.<anonymous> (D:\Projects\playground\tauri3\node_modules\@tauri-apps\api\window.cjs:1:10927)
    at Module._compile (node:internal/modules/cjs/loader:1126:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1180:10)
    at Module.load (node:internal/modules/cjs/loader:1004:32)
    at Function.Module._load (node:internal/modules/cjs/loader:839:12)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:170:29)
    at ModuleJob.run (node:internal/modules/esm/module_job:193:25)
    at async Promise.all (index 0)
    at async ESMLoader.import (node:internal/modules/esm/loader:533:24)
    at async nodeImport (file:///D:/Projects/playground/tauri3/node_modules/vite/dist/node/chunks/dep-67e7f8ab.js:53137:21)
10:28:27 PM [vite-plugin-svelte] ssr compile done.
```

This is a known and expected SvelteKit response which happens when SSR is not disabled properly.

So... is SvelteKit's SSR disabled properly in the template?  Or perhaps this is a problem with @sveltejs/adapter-static?  Is this worth creating an issue for, or must we wait for SvelteKit 1.0?

Again, thank you!

**Comment by melMass** [comment]
AFAIK this is expected behavior?
Even with SSR off the code is "converted" by svelte compiler.

So I would just move this in onMount

I'm maybe wrong as things moves fast in SK, I contributed this when they announced that there should not be major changes to the API for 1.0

**Comment by FireCoding** [comment]
The issue is with the `import` line, and you can't move an `import` into onMount (doesn't help).  This underlying issue is that the ssr compile step is being performed when it shouldn't be.

I've been looking into it and found that the SK team is still making changes (as of just a few days ago) related to builds and prerendering, etc... so it's probably on their side of things.  There might be a minor patch needed to this template once they have it sorted out.  I'll try to keep an eye on it.

**Comment by JDR42** [comment]
@melMass I found out that the new way to disable SSR is to add:

`export const ssr = false;`  in `/src/routes/+layout.ts`

This fixes the issue.  The issue and solution are described [here](https://github.com/sveltejs/kit/tree/master/packages/adapter-static#turn-off-ssr).

It seems to me this should be the default behavior in tauri for the sveltekit template.  Do you think I should create a PR for this?

**Comment by melMass** [comment]
> @melMass I found out that the new way to disable SSR is to add:
> 
> 
> 
> `export const ssr = false;`  in `/src/routes/+layout.ts`

This was already the case at the time (packages/cli/fragments/fragment-svelte-kit-ts/src/routes/+layout.ts) but maybe they changed the way they are triggered IIRC pre-render induced ssr off. And I use this template in a prod app

> Do you think I should create a PR for this?

But yes of course if you think it should be done differently don't hesitate to PR the edit!

**Pull request merged**